arch/posix/src/syscall.rs

```diff
@@ -0,0 +1,539 @@
+            self.write_register(Regs::TRAPNO, kernel_context.read_register(Regs::TRAPNO));
+            self.write_register(Regs::OLDMASK, kernel_context.read_register(Regs::OLDMASK));
+            self.write_register(Regs::CR2, kernel_context.read_register(Regs::CR2));
+        }
+    }
+}
+
+static mut KERNEL_CONTEXT: PosixContext = PosixContext::for_kernel();
+pub static mut CURRENT_PROCESS: PosixStoredState = PosixStoredState::empty_user();
+
+#[no_mangle]
+#[inline(never)]
+pub unsafe fn switch_context_kernel(user_state: &mut PosixStoredState, context: &mut PosixContext) {
+    // save application context
+    std::ptr::copy_nonoverlapping(
+        context as *mut PosixContext as *mut u8,
+        &mut user_state.context as *mut PosixContext as *mut u8,
+        std::mem::size_of::<PosixContext>(),
+    );
+    // restore kernel context
+    // offset the return address
+    std::ptr::copy_nonoverlapping(
+        &mut KERNEL_CONTEXT as *mut PosixContext as *mut u8,
+        context as *mut PosixContext as *mut u8,
+        std::mem::size_of::<PosixContext>(),
+    );
+}
+
+// used for debugging
+#[no_mangle]
+#[inline(never)]
+unsafe fn switched_context_kernel() {
+    llvm_asm! ("nop"::::"volatile");
+    write_volatile(&mut RUNNING_KERNEL, true);
+}
+
+// used for debugging
+#[no_mangle]
+#[inline(never)]
+unsafe fn switched_context_user() {
+    llvm_asm! ("nop"::::"volatile");
+    write_volatile(&mut RUNNING_KERNEL, false);
+}
+
+#[no_mangle]
+#[inline(never)]
+pub unsafe fn switch_context_user(user_state: &mut PosixStoredState, context: &mut PosixContext) {
+    // save kernel context
+    context.write_register(
+        Regs::pc(),
+        context.read_register(Regs::pc()) + POSIX_NEXT_KERNEL_PC,
+    );
+    std::ptr::copy_nonoverlapping(
+        context as *mut PosixContext as *mut u8,
+        &mut KERNEL_CONTEXT as *mut PosixContext as *mut u8,
+        std::mem::size_of::<PosixContext>(),
+    );
+    // restore application context
+    // offset the return address
+
+    std::ptr::copy_nonoverlapping(
+        &mut user_state.context as *mut PosixContext as *mut u8,
+        context as *mut PosixContext as *mut u8,
+        std::mem::size_of::<PosixContext>(),
+    );
+
+    if user_state.first_run {
+        std::ptr::copy_nonoverlapping(
+            &KERNEL_CONTEXT._reserved.as_ptr(),
+            &mut context._reserved.as_ptr(),
+            KERNEL_CONTEXT._reserved.len(),
+        );
+
+        context.first_run(&KERNEL_CONTEXT);
+    }
+
+    context.write_register(Regs::pc(), user_state.sys_pc);
+    context.write_register(Regs::sp(), user_state.stack_pointer);
+}
+
+#[inline(never)]
+#[no_mangle]
+fn switch_to_user() {
+    unsafe {
+        #[cfg(all(target_arch = "x86_64", target_os = "linux"))]
+        llvm_asm!(
+            "
+            mov rax, 0x515CA11F
+            mov [0], rax
+            "
+            : : : "rax" : "volatile", "intel"
+        )
+    }
```

Comment:
This seems to be the only part that is actually x86 specific (from a quick skim). Is there anyway we could change this to not call assembly code?

If this wasn't x86 specific we could run Tock on any architecture. So we could test RISC-V Tock and RISC-V applications as a userspace application. That would be really cool.

It would not use the RISC-V or ARM arch or board, but it would allow us to test syscalls from applicaions to Tock while running on Linux.

Reply:
I'm not sure how, the value has to be placed in a register that is architecture dependent. Mayer adding several cfgs?

Reply:
Yeah, I see what you mean.

I think the current context switch is a little clunky. It seems best to try and convert it to not be architecture specific at all, instead just using POSIX signals.

Couldn't the app be a signal handler and the kernel return another signal handler (alarm signal for example)?

Reply:
There are some limitations when running code inside a signal handler like:
- unable to use some POSIX API (like print)
- unable to preempt a process (or simulate interrupts)

Using this, it will still be platform dependent.

My suggestions is to use #[cfg(....)] and have some code differently. The number of code lines should not be significant.

Reply:
Your current implementation also requires changes to userspace though.

What we want is that the same ARM or RISC-V app that runs on a board will run on a POSIX Tock kernel (on the same architecture).

Reply:
Porting the POSIX arch to ARM and RISC-V should not require any change in the user space (except svc, as that would trigger the guest os system calls).

The changes made to the user space right now are only due to relocation issues for x86.

Reply:
It might not be possible to run the same say RISC-V app binary on a riscv microcontroller and a riscv linux machine. Maybe there is a way to get that to work, but it's probably not an important goal at this point.

However, the "posix" arch in tock shouldn't be x86 specific. Can we save the syscall state somewhere else? Would just a global variable work? Or sent with a signal somehow?

Reply:
> It might not be possible to run the same say RISC-V app binary on a riscv microcontroller and a riscv linux machine. Maybe there is a way to get that to work, but it's probably not an important goal at this point.
> 
> However, the "posix" arch in tock shouldn't be x86 specific. Can we save the syscall state somewhere else? Would just a global variable work? Or sent with a signal somehow?

The state comes within the signal function parameters, saving and restoring it is not a problem. The problem is that it needs to be changed, to "jump" over the actual fault (PC must be incremented). Additionally, system call parameters are in the registers (even tough this might be solved by using the stack).

Another issue that is different is function parameter passing, as this is different from x86, x86_64, arm, ...

What about exposing the architecture as a feature? It would still use #[cfg(...)] ...